runtime/src/main/java/org/corfudb/runtime/object/StreamViewSMRAdapter.java

```diff
@@ -41,11 +44,27 @@ public StreamViewSMRAdapter(CorfuRuntime runtime,
         this.streamView = streamView;
     }
 
+    private List<SMREntry> dataAndCheckpointMapper(ILogData logData) {
+        if (logData.hasCheckpointMetadata()) {
+            // This is a CHECKPOINT record.  Extract the SMREntries, if any.
+            CheckpointEntry cp = (CheckpointEntry) logData.getPayload(runtime);
+            if (cp.getSmrEntries() != null && cp.getSmrEntries().length > 0) {
+                List<SMREntry> consumables = new ArrayList<>();
+                Collections.addAll(consumables, cp.getSmrEntries());
+                return consumables;
+            } else {
+                return (List<SMREntry>) Collections.EMPTY_LIST;
+            }
+        } else {
+            return ((ISMRConsumable) logData.getPayload(runtime)).getSMRUpdates(streamView.getID());
+        }
+    }
+
     public List<SMREntry> remainingUpTo(long maxGlobal) {
         return streamView.remainingUpTo(maxGlobal).stream()
                 .filter(m -> m.getType() == DataType.DATA)
-                .filter(m -> m.getPayload(runtime) instanceof ISMRConsumable)
-                .map(logData -> ((ISMRConsumable)logData.getPayload(runtime)).getSMRUpdates(streamView.getID()))
+                .filter(m -> m.getPayload(runtime) instanceof ISMRConsumable || m.hasCheckpointMetadata())
```

Comment:
shouldn't this be done at the stream layer rather than the SMRStream layer? (i.e., the stream layer converts checkpoints to SMREntries which get fed into this adapter?)